Testcase not found for test result - pytest on pytz

1. git clone https://github.com/stub42/pytz
2. open workspace
3. use this pythonsettings.json
```
{
  "TestFramework": "pytest",
  "SearchPaths": [ "src" ]
}
```

4. run test:
`test_lazy::src\pytz\tests\test_lazy.py::LazyListTestCase::test_append`

Result:
```
[8/16/2019 1:45:32.831 PM Error] Testcase not found for test result: <testcase classname="pytz.tests.test_lazy.LazyListTestCase" file="pytz\tests\test_lazy.py" line="91" name="test_append" time="0.002"></testcase>
```

Full output:
```
[8/16/2019 1:45:20.541 PM Informational] Store opened in 0.000 sec.
[8/16/2019 1:45:20.833 PM Informational] ---------- Discovery started ----------
[8/16/2019 1:45:21.723 PM Informational] Starting pytest discovery for the pytz workspace. This can take up to 60 seconds.
[8/16/2019 1:45:21.724 PM Informational] cd C:\Users\huvalo\source\repos\pytz
[8/16/2019 1:45:21.725 PM Informational] set PYTHONPATH=C:\Users\huvalo\source\repos\pytz;C:\Users\huvalo\source\repos\pytz\;C:\Users\huvalo\source\repos\pytz\src\
[8/16/2019 1:45:21.725 PM Informational] C:\Program Files\Python36\python.exe C:\USERS\HUVALO\APPDATA\LOCAL\MICROSOFT\VISUALSTUDIO\16.0_D1270DCBEXP\EXTENSIONS\MICROSOFT CORPORATION\PYTHON\16.0.0\PythonFiles\testing_tools\run_adapter.py discover pytest --output-file C:\Users\huvalo\AppData\Local\Temp\tmp30C9.tmp -- --cache-clear
[8/16/2019 1:45:22.424 PM Error] equivalent command: C:\Program Files\Python36\python.exe -m pytest --collect-only --cache-clear
pytest discovery failed (exit code 2)
============================= test session starts =============================
platform win32 -- Python 3.6.8, pytest-5.1.0, py-1.8.0, pluggy-0.12.0
rootdir: C:\Users\huvalo\source\repos\pytz
collected 42 items / 1 errors / 41 selected
<Module test_zdump.py>
  <Function test_suite>
<Module src/pytz/tests/test_docs.py>
  <UnitTestCase DocumentationTestCase>
    <TestCaseFunction test_readme_encoding>
  <Function test_suite>
<Module src/pytz/tests/test_lazy.py>
  <UnitTestCase LazyListTestCase>
    <TestCaseFunction test_append>
    <TestCaseFunction test_binary_ops>
    <TestCaseFunction test_bool>
    <TestCaseFunction test_callable>
    <TestCaseFunction test_count>
    <TestCaseFunction test_delitem>
    <TestCaseFunction test_delslice>
    <TestCaseFunction test_extend>
    <TestCaseFunction test_getitem>
    <TestCaseFunction test_getslice>
    <TestCaseFunction test_hash>
    <TestCaseFunction test_iadd>
    <TestCaseFunction test_index>
    <TestCaseFunction test_insert>
    <TestCaseFunction test_isinstance>
    <TestCaseFunction test_iter>
    <TestCaseFunction test_pop>
    <TestCaseFunction test_remove>
    <TestCaseFunction test_reverse>
    <TestCaseFunction test_reversed>
    <TestCaseFunction test_setitem>
    <TestCaseFunction test_setslice>
    <TestCaseFunction test_sort>
    <TestCaseFunction test_sorted>
    <TestCaseFunction test_unary_ops>
  <UnitTestCase LazySetTestCase>
    <TestCaseFunction test_add>
    <TestCaseFunction test_binary_ops>
    <TestCaseFunction test_bool>
    <TestCaseFunction test_callable>
    <TestCaseFunction test_clear>
    <TestCaseFunction test_copy>
    <TestCaseFunction test_discard>
    <TestCaseFunction test_hash>
    <TestCaseFunction test_iops>
    <TestCaseFunction test_isinstance>
    <TestCaseFunction test_method_ops>
    <TestCaseFunction test_pop>
    <TestCaseFunction test_remove>
    <TestCaseFunction test_unary_ops>
=================================== ERRORS ====================================
_______________ ERROR collecting src/pytz/tests/test_tzinfo.py ________________
src\pytz\tests\test_tzinfo.py:40: in <module>
    GMT = pytz.timezone('GMT')
src\pytz\__init__.py:172: in timezone
    zone = _case_insensitive_zone_lookup(_unmunge_zone(zone))
src\pytz\__init__.py:198: in _case_insensitive_zone_lookup
    _all_timezones_lower_to_standard = dict((tz.lower(), tz) for tz in all_timezones)  # noqa
E   NameError: name 'all_timezones' is not defined
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!
============================== 1 error in 0.23s ===============================

[8/16/2019 1:45:22.502 PM Informational] ========== Discovery finished: 42 tests found (0:00:01.6662313) ==========
[8/16/2019 1:45:30.151 PM Informational] ---------- Discovery started ----------
[8/16/2019 1:45:30.762 PM Informational] Starting pytest discovery for the pytz workspace. This can take up to 60 seconds.
[8/16/2019 1:45:30.772 PM Informational] cd C:\Users\huvalo\source\repos\pytz
[8/16/2019 1:45:30.774 PM Informational] set PYTHONPATH=C:\Users\huvalo\source\repos\pytz;C:\Users\huvalo\source\repos\pytz\;C:\Users\huvalo\source\repos\pytz\src\
[8/16/2019 1:45:30.775 PM Informational] C:\Program Files\Python36\python.exe C:\USERS\HUVALO\APPDATA\LOCAL\MICROSOFT\VISUALSTUDIO\16.0_D1270DCBEXP\EXTENSIONS\MICROSOFT CORPORATION\PYTHON\16.0.0\PythonFiles\testing_tools\run_adapter.py discover pytest --output-file C:\Users\huvalo\AppData\Local\Temp\tmp5420.tmp -- --cache-clear
[8/16/2019 1:45:31.321 PM Error] equivalent command: C:\Program Files\Python36\python.exe -m pytest --collect-only --cache-clear
pytest discovery failed (exit code 2)
============================= test session starts =============================
platform win32 -- Python 3.6.8, pytest-5.1.0, py-1.8.0, pluggy-0.12.0
rootdir: C:\Users\huvalo\source\repos\pytz
collected 42 items / 1 errors / 41 selected
<Module test_zdump.py>
  <Function test_suite>
<Module src/pytz/tests/test_docs.py>
  <UnitTestCase DocumentationTestCase>
    <TestCaseFunction test_readme_encoding>
  <Function test_suite>
<Module src/pytz/tests/test_lazy.py>
  <UnitTestCase LazyListTestCase>
    <TestCaseFunction test_append>
    <TestCaseFunction test_binary_ops>
    <TestCaseFunction test_bool>
    <TestCaseFunction test_callable>
    <TestCaseFunction test_count>
    <TestCaseFunction test_delitem>
    <TestCaseFunction test_delslice>
    <TestCaseFunction test_extend>
    <TestCaseFunction test_getitem>
    <TestCaseFunction test_getslice>
    <TestCaseFunction test_hash>
    <TestCaseFunction test_iadd>
    <TestCaseFunction test_index>
    <TestCaseFunction test_insert>
    <TestCaseFunction test_isinstance>
    <TestCaseFunction test_iter>
    <TestCaseFunction test_pop>
    <TestCaseFunction test_remove>
    <TestCaseFunction test_reverse>
    <TestCaseFunction test_reversed>
    <TestCaseFunction test_setitem>
    <TestCaseFunction test_setslice>
    <TestCaseFunction test_sort>
    <TestCaseFunction test_sorted>
    <TestCaseFunction test_unary_ops>
  <UnitTestCase LazySetTestCase>
    <TestCaseFunction test_add>
    <TestCaseFunction test_binary_ops>
    <TestCaseFunction test_bool>
    <TestCaseFunction test_callable>
    <TestCaseFunction test_clear>
    <TestCaseFunction test_copy>
    <TestCaseFunction test_discard>
    <TestCaseFunction test_hash>
    <TestCaseFunction test_iops>
    <TestCaseFunction test_isinstance>
    <TestCaseFunction test_method_ops>
    <TestCaseFunction test_pop>
    <TestCaseFunction test_remove>
    <TestCaseFunction test_unary_ops>
=================================== ERRORS ====================================
_______________ ERROR collecting src/pytz/tests/test_tzinfo.py ________________
src\pytz\tests\test_tzinfo.py:40: in <module>
    GMT = pytz.timezone('GMT')
src\pytz\__init__.py:172: in timezone
    zone = _case_insensitive_zone_lookup(_unmunge_zone(zone))
src\pytz\__init__.py:198: in _case_insensitive_zone_lookup
    _all_timezones_lower_to_standard = dict((tz.lower(), tz) for tz in all_timezones)  # noqa
E   NameError: name 'all_timezones' is not defined
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!
============================== 1 error in 0.17s ===============================

[8/16/2019 1:45:31.399 PM Informational] ========== Discovery finished: 42 tests found (0:00:01.2444905) ==========
[8/16/2019 1:45:31.406 PM Informational] ---------- Run started ----------
[8/16/2019 1:45:32.410 PM Informational] cd C:\Users\huvalo\source\repos\pytz
[8/16/2019 1:45:32.411 PM Informational] set PYTHONPATH=C:\Users\huvalo\source\repos\pytz;C:\Users\huvalo\source\repos\pytz\;C:\Users\huvalo\source\repos\pytz\src\
[8/16/2019 1:45:32.412 PM Informational] "C:\Program Files\Python36\python.exe" "C:\USERS\HUVALO\APPDATA\LOCAL\MICROSOFT\VISUALSTUDIO\16.0_D1270DCBEXP\EXTENSIONS\MICROSOFT CORPORATION\PYTHON\16.0.0\testlauncher.py" C:\Users\huvalo\source\repos\pytz pytest "" 0 "C:\USERS\HUVALO\APPDATA\LOCAL\MICROSOFT\VISUALSTUDIO\16.0_D1270DCBEXP\EXTENSIONS\MICROSOFT CORPORATION\PYTHON\16.0.0\Packages" dummyfilename .\src\pytz\tests\test_lazy.py::LazyListTestCase::test_append --junitxml=C:\Users\huvalo\source\repos\pytz\TestResults\junitresults_629ecf2f-4b02-402d-ad8c-4cb89c1f74e7.xml
[8/16/2019 1:45:32.784 PM Informational] ============================= test session starts =============================
[8/16/2019 1:45:32.784 PM Informational] platform win32 -- Python 3.6.8, pytest-5.1.0, py-1.8.0, pluggy-0.12.0
[8/16/2019 1:45:32.785 PM Informational] rootdir: C:\Users\huvalo\source\repos\pytz\src
[8/16/2019 1:45:32.795 PM Informational] collected 1 item
[8/16/2019 1:45:32.809 PM Informational] src\pytz\tests\test_lazy.py .                                            [100%]
[8/16/2019 1:45:32.809 PM Informational] - generated xml file: C:\Users\huvalo\source\repos\pytz\TestResults\junitresults_629ecf2f-4b02-402d-ad8c-4cb89c1f74e7.xml -
[8/16/2019 1:45:32.810 PM Informational] ============================== 1 passed in 0.03s ==============================
[8/16/2019 1:45:32.831 PM Error] Testcase not found for test result: <testcase classname="pytz.tests.test_lazy.LazyListTestCase" file="pytz\tests\test_lazy.py" line="91" name="test_append" time="0.002"></testcase>
[8/16/2019 1:45:32.895 PM Informational] ========== Run finished: 1 tests run (0:00:01.4853541) ==========

```

junit xml:
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
   <testsuite errors="0" failures="0" hostname="HUVALODESKTOP" name="pytest" skipped="0" tests="1" time="0.022" timestamp="2019-08-16T13:45:32.778426">
      <testcase classname="pytz.tests.test_lazy.LazyListTestCase" file="pytz\tests\test_lazy.py" line="91" name="test_append" time="0.002" />
   </testsuite>
</testsuites>
```

![image](https://user-images.githubusercontent.com/1696845/63197522-e7b4f180-c02c-11e9-9c4e-a4461cd05de4.png)

--- a/Python/Product/TestAdapter.Executor/Pytest/TestResultParser.cs
+++ b/Python/Product/TestAdapter.Executor/Pytest/TestResultParser.cs
@@ -67,7 +67,7 @@ namespace Microsoft.PythonTools.TestAdapter.Pytest {
         /// <returns></returns>
         internal static string CreatePytestId(string filename, string classname, string funcname) {
             var classNameWithoutFilename = String.Empty;
-            var filenamePortion = filename.Replace(".py", "").Replace("\\", ".");
+            var filenamePortion = Path.ChangeExtension(filename, null).Replace("\\", ".");
             if (filenamePortion.Length < classname.Length) {
                 classNameWithoutFilename = classname.Substring(filenamePortion.Length + 1);
             }

--- a/Python/Product/TestAdapter.Executor/Pytest/TestResultParser.cs
+++ b/Python/Product/TestAdapter.Executor/Pytest/TestResultParser.cs
@@ -66,9 +66,13 @@ namespace Microsoft.PythonTools.TestAdapter.Pytest {
         /// <param name="funcname"></param>
         /// <returns></returns>
         internal static string CreatePytestId(string filename, string classname, string funcname) {
-            var classIndex = classname.LastIndexOf(".");
-            if (classIndex >= 0 && classIndex < (classname.Length - 1)) {
-                var classNameWithoutFilename = classname.Substring(classIndex + 1);
+            var classNameWithoutFilename = String.Empty;
+            var filenamePortion = filename.Replace(".py", "").Replace("\\", ".");
+            if (filenamePortion.Length < classname.Length) {
+                classNameWithoutFilename = classname.Substring(filenamePortion.Length + 1);
+            }
+
+            if (!String.IsNullOrEmpty(classNameWithoutFilename)) {
                 return $".\\{filename}::{classNameWithoutFilename}::{funcname}";
             } else {
                 return $".\\{filename}::{funcname}";

--- a/Python/Tests/TestAdapterTests/TestResultParserTests.cs
+++ b/Python/Tests/TestAdapterTests/TestResultParserTests.cs
@@ -39,6 +39,14 @@ namespace TestAdapterTests {
         }
 
         [TestMethod]
+        public void CreatePytestId_GlobalFuncRelative() {
+            Assert.AreEqual(
+                ".\\tests\\unit\\test_statistics.py::test_key_creation",
+                TestResultParser.CreatePytestId("tests\\unit\\test_statistics.py", "tests.unit.test_statistics", "test_key_creation")
+            );
+        }
+
+        [TestMethod]
         public void CreatePytestId_ClassFuncWithRelativeFilename() {
             Assert.AreEqual(
                 ".\\package1\\packageA\\test1.py::Test_test1::test_A", 

